lib/internal/streams/web.js

```diff
@@ -0,0 +1,4394 @@
+  Uint8Array,
+} = primordials;
+
+const {
+  codes: {
+    ERR_ILLEGAL_CONSTRUCTOR,
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_STATE,
+    ERR_INVALID_THIS,
+  },
+} = require('internal/errors');
+
+const {
+  DOMException,
+} = internalBinding('messaging');
+
+const {
+  copyArrayBuffer,
+  detachArrayBuffer
+} = internalBinding('buffer');
+
+const {
+  isArrayBufferView,
+  isDataView,
+  isPromise,
+} = require('util/types');
+
+const {
+  createDeferredPromise,
+  emitExperimentalWarning,
+  customInspectSymbol: kInspect,
+} = require('internal/util');
+
+const {
+  inspect,
+} = require('util');
+
+const {
+  serialize,
+  deserialize,
+} = require('v8');
+
+const {
+  validateObject,
+} = require('internal/validators');
+
+const {
+  kAborted,
+} = require('internal/abort_controller');
+
+const {
+  MessageChannel,
+} = require('internal/worker/io');
+
+const {
+  kDeserialize,
+  kTransfer,
+  kTransferList,
+  JSTransferable,
+} = require('internal/worker/js_transferable');
+
+const {
+  getPromiseDetails,
+  kPending,
+} = internalBinding('util');
+
+const {
+  queueMicrotask,
+} = require('internal/process/task_queues');
+
+const assert = require('internal/assert');
+
+emitExperimentalWarning('stream/web');
+
+const kAbort = Symbol('kAbort');
+const kCancel = Symbol('kCancel');
+const kClose = Symbol('kClose');
+const kCloseSentinel = Symbol('kCloseSentinel');
+const kChunk = Symbol('kChunk');
+const kError = Symbol('kError');
+const kPull = Symbol('kPull');
+const kState = Symbol('kState');
+const kType = Symbol('kType');
+
+const AsyncIteratorPrototype = ObjectGetPrototypeOf(
+  ObjectGetPrototypeOf(async function* () {}).prototype);
+
+function extractHighWaterMark(value, defaultHWM) {
+  if (value === undefined) return defaultHWM;
+  if (typeof value !== 'number' ||
+      NumberIsNaN(value) ||
+      value < 0)
+    throw new ERR_INVALID_ARG_VALUE.RangeError('strategy.highWaterMark', value);
+  return value;
+}
+
+function extractSizeAlgorithm(size) {
+  if (size === undefined) return () => 1;
+  if (typeof size !== 'function')
+    throw new ERR_INVALID_ARG_VALUE('strategy.size', size);
+  return size;
+}
+
+function customInspect(depth, options, name, data) {
+  if (depth < 0)
+    return this;
+
+  const opts = {
+    ...options,
+    depth: options.depth == null ? null : options.depth - 1
+  };
+
+  return `${name} ${inspect(data, opts)}`;
+}
+
+// These are defensive to work around the possibility that
+// the buffer, byteLength, and byteOffset properties on
+// ArrayBuffer and ArrayBufferView's may have been tampered with.
+
+function ArrayBufferViewGetBuffer(view) {
+  return ReflectGet(view.constructor.prototype, 'buffer', view);
+}
+
+function ArrayBufferViewGetByteLength(view) {
+  return ReflectGet(view.constructor.prototype, 'byteLength', view);
+}
+
+function ArrayBufferViewGetByteOffset(view) {
+  return ReflectGet(view.constructor.prototype, 'byteOffset', view);
+}
```

Comment:
unfortunately, the `constructor` property is never reliable :-/ these might need to use internal brand-checking methods to determine the appropriate getter to use.

Reply:
Yeah, for now I'm going to leave these as is and focus on the other bits. We can fix this piece up before it lands (hopefully) or before it progresses out of experimental